### PR TITLE
Enhancements to subscriber lists report

### PR DESCRIPTION
**Pull request by benthorner**
https://trello.com/c/PRXBoZBu/576-create-a-one-stop-data-shop-report-task-for-all-subscriber-lists

This adds a couple of new filtering capabilities:

- Filtering by a comma-separated list of slugs
- Filtering by list tags (as a string)
- Filtering by list links (as a string)

Please see the commits for more details.

### CODE REVIEW

**Review by kevindew — COMMENTED** [review]
Looks good - nice cleanup. Just a few minor comments.

File: lib/reports/subscriber_lists_report.rb [hunk 1]
```diff
@@ -12,30 +12,42 @@ class Reports::SubscriberListsReport
                    matched_content_changes_for_date
                    matched_messages_for_date].freeze
 
-  def initialize(date)
+  def initialize(date, slugs: nil)
     @date = Time.zone.parse(date)
+    @slugs = slugs ? slugs.split(",") : nil
```
[the remark this answers is not in the document]
Comment (benthorner) [reply lib/reports/subscriber_lists_report.rb]: Oh yeah, good idea - I've changed it to assume an empty string.

File: lib/reports/subscriber_lists_report.rb [hunk 2]
```diff
@@ -32,6 +33,7 @@ def call
   def lists_to_report
     scope = SubscriberList.where("created_at < ?", date.end_of_day)
     scope = scope.where(slug: slugs) if slugs
+    scope = scope.where("tags::text like ?", "%#{tags_pattern}%") if tags_pattern
```
[the remark this answers is not in the document]
Comment (benthorner) [reply lib/reports/subscriber_lists_report.rb:37]: Hmm...I didn't think about that. Let's wait and see if it's actually required in practice, though.

I did check performance, as I'd assumed the search would be quite slow for a non-indexed field. But it's actually really fast for the couple of test queries I did. And yes, the fact that it's a rake task means it's not critical anyway.

File: spec/lib/reports/subscriber_lists_report_spec.rb [hunk 3]
```diff
@@ -35,6 +35,16 @@
     expect(output.lines.count). to eq 2
   end
 
+  it "can filter based on list tags (as a string)" do
+    create(:subscriber_list, tags: {})
+
+    output = described_class.new("2020-06-15", tags_pattern: "road").call
+    expect(output.lines.count). to eq 2
```
[the remark this answers is not in the document]
Comment (benthorner) [reply spec/lib/reports/subscriber_lists_report_spec.rb]: Oops - removed 👍.

**Comment by benthorner** [comment]
Thanks @kevindew - I'm pleased we can get rid of the specific reports!

This should be ready for another look now.

File: lib/reports/subscriber_lists_report.rb [hunk 4]
```diff
@@ -12,32 +12,50 @@ class Reports::SubscriberListsReport
...
+  end
+
   def validate_date
     raise "Invalid date" if date.blank?
     raise "Date must be in the past" if date >= Time.zone.today
     raise "Date must be within a year old" if date <= 1.year.ago
   end
 
+  def validate_slugs
+    return unless slugs
```
Comment (kevindew) [inline lib/reports/subscriber_lists_report.rb:49]: This looks like it missed a tweak in the conversion to string. Should it be `if slugs.empty?`?

Comment (benthorner) [reply lib/reports/subscriber_lists_report.rb]: Oh yeah. I realise we don't need that line at all anymore - removed!

**Review by kevindew — APPROVED** [review]

**Pull request merged**